Source/JavaScriptCore/runtime/WaiterListsManager.h

```diff
@@ -0,0 +1,330 @@
+ * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+ * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+#pragma once
+
+#include "JSGlobalObject.h"
+#include "JSPromise.h"
+#include "ObjectConstructor.h"
+#include "runtime/DeferredWorkTimer.h"
+#include <wtf/DataLog.h>
+#include <wtf/NeverDestroyed.h>
+#include <wtf/RawPointer.h>
+
+namespace JSC {
+
+namespace WaiterListsManagerInternal {
+static constexpr bool verbose = false;
+}
+
+struct Waiter {
+    WTF_MAKE_FAST_ALLOCATED;
+
+public:
+    Waiter(JSPromise* promise)
+        : m_isAsync(true)
+    {
+        VM& vm = promise->globalObject()->vm();
+        vm.apiLock().lock();
+        data.ticket = vm.deferredWorkTimer->addPendingWork(vm, promise, { });
+        vm.apiLock().unlock();
+    }
+
+    Waiter(VM* vm)
+        : m_isAsync(false)
+    {
+        data.vm = vm;
+    }
+
+    bool isAsync() const
+    {
+        return m_isAsync;
+    }
+
+    VM* getVM() const
+    {
+        return m_isAsync ? &getPromise()->globalObject()->vm() : data.vm;
+    }
+
+    JSPromise* getPromise() const
+    {
+        RELEASE_ASSERT(m_isAsync);
+        return jsCast<JSPromise*>(data.ticket->target());
+    }
+
+    DeferredWorkTimer::Ticket getTicket() const
+    {
+        RELEASE_ASSERT(m_isAsync);
+        return data.ticket;
+    }
+
+    void dump(PrintStream& out) const
+    {
+        if (m_isAsync)
+            out.print("{promise: ", RawPointer(getPromise()), "}");
+        else
+            out.print("{vm: ", RawPointer(data.vm), "}");
+    }
+
+private:
+    bool m_isAsync { false };
+    union {
+        DeferredWorkTimer::Ticket ticket;
+        VM* vm;
+    } data;
+};
+
+struct WaiterList {
+    WTF_MAKE_FAST_ALLOCATED;
+
+public:
+    void enqueue(JSPromise* promise)
+    {
+        queue.append(makeUnique<Waiter>(promise));
+    }
+
+    void enqueue(VM* promise)
+    {
+        queue.append(makeUnique<Waiter>(promise));
+    }
+
+    std::unique_ptr<Waiter> dequeue()
+    {
+        return queue.takeFirst();
+    }
+
+    template<typename Functor>
+    std::unique_ptr<Waiter> takeFirst(const Functor& functor)
+    {
+        return queue.takeFirst(functor);
+    }
+
+    bool isEmpty()
+    {
+        return queue.isEmpty();
+    }
+
+    void dump(PrintStream& out) const
+    {
+        out.print("queue: [");
+        for (auto& waiter : queue) {
+            waiter->dump(out);
+            out.print(" ");
+        }
+        out.print("]");
+    }
+
+    mutable Mutex lock;
+
+private:
+    Deque<std::unique_ptr<Waiter>> queue;
+};
+
+class WaiterListsManager {
+    WTF_MAKE_FAST_ALLOCATED;
+
+public:
+    static WaiterListsManager& singleton()
+    {
+        static LazyNeverDestroyed<WaiterListsManager> manager;
+        static std::once_flag onceKey;
+        std::call_once(onceKey, [&] {
+            manager.construct();
+        });
+        return manager;
+    }
```

Comment:
This should be defined in cpp file. Otherwise, this `static` variable becomes copied in multiple translation units.